File: server/cpp_server/src/dag/dag.cc
/*
 * Copyright 2022 netease. All rights reserved.
 * Author zhaochaochao@corp.netease.com
 * Date   2023/02/15
 * Brief  Dag of inference.
 */

#include "dag/dag.h"

#include "logger/logger.h"

namespace netease::grps {
void SequentialDag::BuildDag(const std::vector<GlobalConfig::InferenceConfig::NodeConfig>& node_configs,
                             const std::unordered_map<std::string, Model>& models) {
  LOG4(INFO, "Build sequential dag: " << name_);
  for (const auto& node : node_configs) {
    if (node.type == "model") {
      auto model = models.find(node.model);
      if (model == models.end()) {
        LOG4(ERROR, "Model not found: " << node.model);
        throw InferDagException("Model not found: " + node.model);
      }
      sequence_.emplace_back(std::make_shared<ModelNode>(node.name, model->second.inferer_, model->second.converter_,
                                                         model->second.batcher_));
    } else {
      LOG4(ERROR, "Unknown node type: " << node.type);
      throw InferDagException("Unknown node type: " + node.type);
    }
  }
  std::stringstream ss;
  for (const auto& node : sequence_) {
    ss << node->name() << " -> ";
  }
  LOG4(INFO, "Build sequential dag successfully, sequence is: " << ss.str());
}

void SequentialDag::Infer(const ::grps::protos::v1::GrpsMessage& input,
                          ::grps::protos::v1::GrpsMessage& output,
                          GrpsContext& ctx) {
#ifdef GRPS_DEBUG
  LOG4(INFO, "Infer sequential dag: " << name_);
#endif
  for (int i = 0; i < sequence_.size(); ++i) {
    if (i == 0) {
      sequence_[i]->Process(input, output, ctx);
    } else {
      sequence_[i]->Process(output, output, ctx);
    }
    if (ctx.has_err()) {
      return;
    }
  }
}

void SequentialDag::Infer(const ::grps::protos::v1::GrpsMessage& input,
                          ::grps::protos::v1::GrpsMessage& output,
                          const std::shared_ptr<GrpsContext>& ctx_sp) {
#ifdef GRPS_DEBUG
  LOG4(INFO, "Infer sequential dag: " << name_);
#endif
  for (int i = 0; i < sequence_.size(); ++i) {
    if (i == 0) {
      sequence_[i]->Process(input, output, ctx_sp);
    } else {
      sequence_[i]->Process(output, output, ctx_sp);
    }
    if (ctx_sp->has_err()) {
      return;
    }
  }
}

void GraphDag::BuildDag(const std::vector<GlobalConfig::InferenceConfig::NodeConfig>& node_configs,
                        const std::unordered_map<std::string, Model>& models) {}

void GraphDag::Infer(const ::grps::protos::v1::GrpsMessage& input,
                     ::grps::protos::v1::GrpsMessage& output,
                     GrpsContext& ctx) {}

void GraphDag::Infer(const ::grps::protos::v1::GrpsMessage& input,
                     ::grps::protos::v1::GrpsMessage& output,
                     const std::shared_ptr<GrpsContext>& ctx_sp) {}
} // namespace netease::grps

File: server/py_server/grps_framework/dag/dag.py
# Copyright 2022 netease. All rights reserved.
# Author zhaochaochao@corp.netease.com
# Date   2022/9/13
# Brief  DAG of inference.

from abc import ABCMeta, abstractmethod

from grps_framework.context.context import GrpsContext
from grps_framework.dag.node import ModelNode
from grps_framework.logger.logger import logger


class InferDag(metaclass=ABCMeta):
    """Infer DAG base class."""

    @abstractmethod
    def __init__(self, name):
        self._name = name
        pass

    @abstractmethod
    def build_dag(self, nodes_conf, model_dict):
        pass

    @abstractmethod
    def infer(self, data, context):
        pass

    @property
    def name(self):
        return self._name


class SequentialDag(InferDag):
    """Sequential DAG, which means the models will be executed in sequence."""

    def __init__(self, name):
        super().__init__(name)
        self.__sequence = []
        pass

    def build_dag(self, nodes_conf, model_dict):
        """
        Build dag.
        Args:
            nodes_conf: nodes conf.
            model_dict: model dict with key is model name and val is (model_infer, converter, batcher).

        Returns:
            True if build dag successfully, else False.
        """
        logger.info('Build sequential dag.')
        for node in nodes_conf:
            node_name = node['name']
            node_type = node['type']
            if node_type == 'model':
                node_model = node['model']
                if node_model not in model_dict:
                    raise ValueError('Model {} not found but bound with {} node.'.format(node_model, node_name))
                self.__sequence.append(ModelNode(node_name, *model_dict[node_model]))
            else:
                # TODO: Add splitter and merger and etc node type.
                raise ValueError('Node type {} not supported in sequential dag.'.format(node_type))
        seq_str = ' -> '.join([str(node) for node in self.__sequence])
        logger.info('Build sequential dag successfully, sequence is {}.'.format(seq_str))
        return True

    def infer(self, data, context: GrpsContext):
        """
        Sequential dag infer.
        Args:
            context: grps context
            data: input data.

        Returns:
            output will be request response.
        """
        logger.debug('Sequential dag infer.')
        for node in self.__sequence:
            data = node.process(data, context)
            if context.has_err():
                return
        logger.debug('Sequential dag infer successfully.')
        return data


# TODO: Add graph DAG.
class GraphDag(InferDag):
    """Graph DAG, which means the models will be executed in graph."""

    def __init__(self, name):
        super().__init__(name)
        pass

    def build_dag(self, nodes_conf, model_dict):
        pass

    def infer(self, data):
        pass
